### PR TITLE
Wishlist

### PR DESCRIPTION
**Pull request by SepidehMiralai**

### CODE REVIEW

File: db/migrate/20191118183622_create_line_items.rb [hunk 1]
```diff
@@ -1,19 +1,10 @@
 class CreateLineItems < ActiveRecord::Migration[6.0]
   def change
     create_table :line_items do |t|
-      t.references :product, null: false, foreign_key: true
-      t.belongs_to :cart, null: false, foreign_key: true
+      t.references :product, foreign_key: true
+      t.belongs_to :cart, foreign_key: true
```
Comment (bradsokol) [inline db/migrate/20191118183622_create_line_items.rb]: I'm not sure what you're doing here, but it is not a good idea to change a migration that has already been run. Rails runs each migration only once, so if you change the file after the migration has been run, the changes won't take affect. The best approach is to put the necessary changes in a new migration.

File: app/controllers/wishlists_controller.rb [hunk 2]
```diff
@@ -0,0 +1,88 @@
+class WishlistsController < ApplicationController
+  before_action :set_wishlist, only: [:show, :edit, :update, :destroy]
+
+  # GET /wishlists
+  # GET /wishlists.json
+  def index
+    @wishlists = Wishlist.all
+  end
```
[the remark this answers is not in the document]
Comment (SepidehMiralai) [reply app/controllers/wishlists_controller.rb]: Its used in wishlist view (index.html.erb) with a condition of displaying only the wishlist of current user.

File: app/models/product.rb [hunk 3]
```diff
@@ -1,6 +1,8 @@
 class Product < ApplicationRecord
 
     has_many :line_items
+    has_many :orders, through: :line_items
```
[the remark this answers is not in the document]
Comment (SepidehMiralai) [reply app/models/product.rb]: No, its part of implementing orders

File: app/models/user.rb [hunk 4]
```diff
@@ -0,0 +1,15 @@
+class User < ApplicationRecord
```
[the remark this answers is not in the document]
Comment (SepidehMiralai) [reply app/models/user.rb]: Sure

File: app/models/user.rb [hunk 5]
```diff
@@ -0,0 +1,15 @@
...
+  has_one :wishlist
+  after_destroy :ensure_an_admin_remains 
+  class Error < StandardError
+  end
+  private
+    def ensure_an_admin_remains
+      if User.count.zero?
+        raise Error.new "Can't delete last user"
+      end 
+    end
```
[the remark this answers is not in the document]
Comment (SepidehMiralai) [reply app/models/user.rb]: Thanks for letting me know

File: app/controllers/wishlists_controller.rb (hunk 2 above)
[the remark this answers is not in the document]
Comment (bradsokol) [reply app/controllers/wishlists_controller.rb]: I see what's happening. The filtering for the user is happening in the view [here](https://github.com/SepidehMiralai/sepid_shoppingcart/pull/1/files#diff-96f77b0af03e20a948271254304576d2R14).

In a Rails app, it is better to limit the view to only be concerned with presenting the data to the user. Getting and manipulating the data should happen in the controller. It's [cleaner separation](https://en.wikipedia.org/wiki/Separation_of_concerns) of concerns. 

If not done in one place, every view of the code will have to worry about filtering for the current user. This is already a problem when you get the JSON representation of the data, which shows all wish lists for all users. Try this by creating two users, each with a wish list then going to https://sepid-shoppingcart.myshopify.io/wishlists.json.

**Pull request merged**